samples/python/text_generation/compound_grammar_generation.py

```diff
@@ -131,15 +173,25 @@ def main():
     )
     print("User: ", user_text_2)
     chat_history.append({"role": "user", "content": user_text_2})
-    model_input = tokenizer.apply_chat_template(chat_history, add_generation_prompt=True)
+    model_input = tokenizer.apply_chat_template(chat_history, add_generation_prompt=True, tools=tools)
 
     start_tool_call_tag = SOC.ConstString(r"functools")
-    tools_json = SOC.JSONSchema(tools_to_array_schema(booking_flight_tickets, booking_hotels))
+    tools_json = SOC.JSONSchema(tools_to_array_schema(book_flight_ticket, book_hotel))
     tool_call_grammar = start_tool_call_tag + tools_json  # SOC.Concat(start_tool_call_tag, tools_json)
     generation_config.structured_output_config.structural_tags_config = tool_call_grammar
 
     print("Assistant: ", end="")
-    pipe.generate(model_input, generation_config, streamer=streamer)
+    custom_streamer = CurrentTextParserStreamer(pipe.get_tokenizer(), [IncrementalToolCallParser()])
+    answer = pipe.generate([model_input], generation_config, streamer=custom_streamer)
+    print()
+    
+    print("\nThe following tool calls were generated:")
+    def print_tool_call(tool_call: dict):
+        print(f"{tool_call['name']}({', '.join(f'{key}=\"{value}\"' for key, value in tool_call['arguments'].items())})")
```

Comment:
Better to define this function separately.

Reply:
done